### PR TITLE
Implement current_date scalar function

### LINKED ISSUE #3981: Implement `current_date` Function
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
As we add more full featured date/time support for DataFusion users need to be able to use the current date to calculate things like "all data in the last 30 days" 

**Describe the solution you'd like**
Support the `current_date` postgres function https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-CURRENT

**Describe alternatives you've considered**

**Additional context**
This is a feature request we got from an early (internal) user of InfluxDB IOx

I believe this is a good first issue as you can follow the existing implementation of  `now()` 
- [ ] Physical support: https://github.com/apache/arrow-datafusion/blob/f1322592046a7ea895820286a2ad16d934540d92/datafusion/physical-expr/src/functions.rs#L424-L429) and use the same `query_execution_start_time` field
- [ ] Tests: https://github.com/apache/arrow-datafusion/blob/beeb631d0bb979e9b6826f6cd3eae2d0a24f471e/datafusion/core/tests/sql/timestamp.rs#L417

It would be a good exercise to add a new build in scalar function, physical planing support, and then tests

### PR DESCRIPTION
**Pull request by comphead**
# Which issue does this PR close?

Closes #3981

# Rationale for this change
DataFusion users need to be able to use the current date to calculate things like "all data in the last 30 days"

# What changes are included in this PR?
Scalar current_date function

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/tests/sql/timestamp.rs [hunk 1]
```diff
@@ -1619,3 +1619,35 @@ async fn test_cast_to_timetz_should_not_work() -> Result<()> {
...
+        results[0]
+            .schema()
+            .field_with_name("dt")
+            .unwrap()
+            .data_type()
+            .to_owned(),
+        DataType::Date32
+    );
+
+    let sql = "select case when current_date() = cast(now() as date) then 'OK' else 'FAIL' end result";
```
Comment (alamb) [inline datafusion/core/tests/sql/timestamp.rs:1639]: Nice!

File: datafusion/expr/src/built_in_function.rs [hunk 2]
```diff
@@ -254,6 +258,7 @@ impl BuiltinScalarFunction {
 
             // Stable builtin functions
             BuiltinScalarFunction::Now => Volatility::Stable,
+            BuiltinScalarFunction::CurrentDate => Volatility::Stable,
```
Comment (alamb) [inline datafusion/expr/src/built_in_function.rs:261]: 👍

File: datafusion/expr/src/expr_fn.rs [hunk 3]
```diff
@@ -452,6 +452,14 @@ pub fn now() -> Expr {
     }
 }
 
+/// Returns current UTC date in default format yyyy-MM-dd
```
Comment (alamb) [inline datafusion/expr/src/expr_fn.rs:455]:
```suggestion
/// Returns current UTC date as a [`DataType::Date32`] value
```

File: datafusion/physical-expr/src/functions.rs [hunk 4]
```diff
@@ -427,6 +427,12 @@ pub fn create_physical_fun(
                 execution_props.query_execution_start_time,
             ))
         }
+        BuiltinScalarFunction::CurrentDate => {
+            // bind value for now at plan time
```
Comment (alamb) [inline datafusion/physical-expr/src/functions.rs:431]:
```suggestion
            // bind value for current_date at plan time
```

**Review by alamb — APPROVED** [review]
Looks great to me @comphead  -- thank you!

I left some minor suggestions, but I think this PR could be merged in without changes as well.

**Review by Jimexist — APPROVED** [review]

**Comment by alamb** [comment]
Since this appears to have some other work building behind it and it has two approvals, merging it in!

**Comment by alamb** [comment]
Thanks again @comphead  and @Jimexist

**Pull request merged**